### PR TITLE
convert test_projection to C++

### PR DESCRIPTION
**Pull request by zingale**

### CODE REVIEW

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 1]
```diff
@@ -90,19 +87,45 @@ void Maestro::Evolve() {
...
 
-                init_vel(ARLIM_3D(lo), ARLIM_3D(hi), BL_TO_FORTRAN_3D(vel[mfi]),
-                         ZFILL(dx));
+                Array4<Real> const vel_arr = vel.array(mfi);
+
+                amrex::ParallelFor(tilebox,
+                [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                {
+
+                    Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:105]: prob_lo[2] -> prob_lo[0]

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 2]
```diff
@@ -90,19 +87,45 @@ void Maestro::Evolve() {
...
+                    Real y = (static_cast<Real>(j)+0.5_rt) * dx[1] + prob_lo[1];
+                    Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:107]: Add `#if AMREX_SPACEDIM == 3` around `Real z = ...`

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 3]
```diff
@@ -115,22 +138,74 @@ void Maestro::Evolve() {
...
+                Array4<Real> const vmac_arr = vmac_mf.array(mfi);
+                Array4<Real> const wmac_arr = wmac_mf.array(mfi);
+
+                // x-velocity  (x are edges, y and z are centers)
+
+                amrex::ParallelFor(mfi.nodaltilebox(0),
+                [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                {
+
+                    Real x = static_cast<Real>(i) * dx[0] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:158]: prob_lo[2] -> prob_lo[0]

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 4]
```diff
@@ -115,22 +138,74 @@ void Maestro::Evolve() {
...
+                    Real y = (static_cast<Real>(j)+0.5_rt) * dx[1] + prob_lo[1];
+                    Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:160]: Add `#if AMREX_SPACEDIM == 3` around `Real z = ...`

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 5]
```diff
@@ -115,22 +138,74 @@ void Maestro::Evolve() {
...
-                             ZFILL(dx));
+                });
+
+                // y-velocity  (x and z are centers, y are edges)
+
+                amrex::ParallelFor(mfi.nodaltilebox(1),
+                [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                {
+
+                    Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:177]: prob_lo[2] -> prob_lo[0]

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 6]
```diff
@@ -115,22 +138,74 @@ void Maestro::Evolve() {
...
+                    Real y = static_cast<Real>(j) * dx[1] + prob_lo[1];
+                    Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
+
+#if AMREX_SPACEDIM == 2
+                    vmac_arr(i,j,k) = std::pow(std::sin(M_PI*y), 2) * std::sin(2.0_rt*M_PI*x);
+#else
+
+                    vmac(i,j,k) =
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:185]: vmac -> vmac_arr

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 7]
```diff
@@ -115,22 +138,74 @@ void Maestro::Evolve() {
...
+
+                // y-velocity  (x and z are centers, y are edges)
+
+                amrex::ParallelFor(mfi.nodaltilebox(1),
+                [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                {
+
+                    Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
+                    Real y = static_cast<Real>(j) * dx[1] + prob_lo[1];
+                    Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:179]: Add `#if AMREX_SPACEDIM == 3` around `Real z = ...`

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 8]
```diff
@@ -115,22 +138,74 @@ void Maestro::Evolve() {
...
+                });
+
+#if AMREX_SPACEDIM == 3
+                // z-velocity  (x and y are centers, z are edges)
+
+                amrex::ParallelFor(mfi.nodaltilebox(2),
+                [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                {
+
+                    Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:198]: prob_lo[2] -> prob_lo[0]

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 9]
```diff
@@ -212,19 +295,124 @@ void Maestro::Evolve() {
...
+                    ) {
+
+                    Array4<Real> const gphi_arr = gphi_mf.array(mfi);
+                    Array4<Real> const umid_arr = umid_mf.array(mfi);
+
+                    amrex::ParallelFor(tilebox,
+                    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                    {
+
+                        Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:328]: prob_lo[2] -> prob_lo[0]

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 10]
```diff
@@ -212,19 +295,124 @@ void Maestro::Evolve() {
...
+                        Real y = (static_cast<Real>(j)+0.5_rt) * dx[1] + prob_lo[1];
+                        Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:330]: Add `#if AMREX_SPACEDIM == 3` around `Real z = ...`

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 11]
```diff
@@ -212,19 +295,124 @@ void Maestro::Evolve() {
...
+                    Elixir elix_phi = phi.elixir();
+                    auto phi_arr = phi.array();
+
+                    // now we can explicitly difference phi
+
+                    amrex::ParallelFor(gbox,
+                    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                    {
+
+                        Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:373]: prob_lo[2] -> prob_lo[0]

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 12]
```diff
@@ -212,19 +295,124 @@ void Maestro::Evolve() {
...
+                        Real y = (static_cast<Real>(j)+0.5_rt) * dx[1] + prob_lo[1];
+                        Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:375]: Add `#if AMREX_SPACEDIM == 3` around `Real z = ...`

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 13]
```diff
@@ -212,19 +295,124 @@ void Maestro::Evolve() {
...
+
+                    Array4<Real> const gphi_arr = gphi_mf.array(mfi);
+                    Array4<Real> const umid_arr = umid_mf.array(mfi);
+
+
+                    amrex::ParallelFor(tilebox,
+                    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                    {
+
+                        Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:394]: prob_lo[2] -> prob_lo[0]

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 14]
```diff
@@ -212,19 +295,124 @@ void Maestro::Evolve() {
...
+                        Real y = (static_cast<Real>(j)+0.5_rt) * dx[1] + prob_lo[1];
+                        Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:396]: Add `#if AMREX_SPACEDIM == 3` around `Real z = ...`

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 15]
```diff
@@ -260,31 +448,220 @@ void Maestro::Evolve() {
...
+                    // x-velocity  (x are edges, y and z are centers)
+
+                    Array4<Real> const gphix_arr = gphix_mac_mf.array(mfi);
+                    Array4<Real> const umac_arr = umac_mid_mf.array(mfi);
+
+                    amrex::ParallelFor(mfi.nodaltilebox(0),
+                    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                    {
+
+                        Real x = static_cast<Real>(i) * dx[0] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:487]: prob_lo[2] -> prob_lo[0]

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 16]
```diff
@@ -260,31 +448,220 @@ void Maestro::Evolve() {
...
+                        Real y = (static_cast<Real>(j)+0.5_rt) * dx[1] + prob_lo[1];
+                        Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:489]: Add `#if AMREX_SPACEDIM == 3` around `Real z = ...`

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 17]
```diff
@@ -260,31 +448,220 @@ void Maestro::Evolve() {
...
+                    // y-velocity  (x and z are centers, y are edges)
+
+                    Array4<Real> const gphiy_arr = gphiy_mac_mf.array(mfi);
+                    Array4<Real> const vmac_arr = vmac_mid_mf.array(mfi);
+
+                    amrex::ParallelFor(mfi.nodaltilebox(1),
+                    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                    {
+
+                        Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:509]: prob_lo[2] -> prob_lo[0]

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 18]
```diff
@@ -260,31 +448,220 @@ void Maestro::Evolve() {
...
+                        Real y = static_cast<Real>(j) * dx[1] + prob_lo[1];
+                        Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:511]: Add `#if AMREX_SPACEDIM == 3` around `Real z = ...`

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 19]
```diff
@@ -260,31 +448,220 @@ void Maestro::Evolve() {
...
+                    // z-velocity  (x and y are centers, z are edges)
+
+                    Array4<Real> const gphiz_arr = gphiz_mac_mf.array(mfi);
+                    Array4<Real> const wmac_arr = wmac_mid_mf.array(mfi);
+
+                    amrex::ParallelFor(mfi.nodaltilebox(2),
+                    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                    {
+
+                        Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:533]: prob_lo[2] -> prob_lo[0]

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 20]
```diff
@@ -260,31 +448,220 @@ void Maestro::Evolve() {
...
+                    Elixir elix_phi = phi.elixir();
+                    auto phi_arr = phi.array();
+
+                    // now we can explicitly difference phi
+
+                    amrex::ParallelFor(gbox,
+                    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                    {
+
+                        Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:567]: prob_lo[2] -> prob_lo[0]

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 21]
```diff
@@ -260,31 +448,220 @@ void Maestro::Evolve() {
...
+                        Real y = (static_cast<Real>(j)+0.5_rt) * dx[1] + prob_lo[1];
+                        Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:569]: Add `#if AMREX_SPACEDIM == 3` around `Real z = ...`

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 22]
```diff
@@ -260,31 +448,220 @@ void Maestro::Evolve() {
...
+                    // x-velocity  (x are edges, y and z are centers)
+
+                    Array4<Real> const gphix_arr = gphix_mac_mf.array(mfi);
+                    Array4<Real> const umac_arr = umac_mid_mf.array(mfi);
+
+                    amrex::ParallelFor(mfi.nodaltilebox(0),
+                    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                    {
+
+                        Real x = static_cast<Real>(i) * dx[0] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:595]: prob_lo[2] -> prob_lo[0]

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 23]
```diff
@@ -260,31 +448,220 @@ void Maestro::Evolve() {
...
+                        Real y = (static_cast<Real>(j)+0.5_rt) * dx[1] + prob_lo[1];
+                        Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
```
Comment (doreenfan) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:597]: Add `#if AMREX_SPACEDIM == 3` around `Real z = ...`

**Review by doreenfan — CHANGES_REQUESTED** [review]
There is a typo when computing x: `prob_lo[2]` should be `prob_lo[0]`. 
And z should only be computed when `AMREX_SPACEDIM == 3`.

**Comment by doreenfan** [comment]
We also need to include `maestro.max_step = 1` in every input file because either `max_step` or `stop_time` needs to be given for the code to run.

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp (hunk 1 above)
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:105]:
```suggestion
                    Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[0];
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp (hunk 2 above)
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:107]:
```suggestion
#if (AMREX_SPACEDIM == 3)
                    Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
#endif
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp (hunk 3 above)
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:158]:
```suggestion
                    Real x = static_cast<Real>(i) * dx[0] + prob_lo[0];
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp (hunk 4 above)
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:160]:
```suggestion
#if (AMREX_SPACEDIM==3)
                    Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
#endif
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 24]
```diff
@@ -115,22 +138,74 @@ void Maestro::Evolve() {
...
+#endif
+                });
+
+                // y-velocity  (x and z are centers, y are edges)
+
+                amrex::ParallelFor(mfi.nodaltilebox(1),
+                [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                {
+
+                    Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:177]:
```suggestion
                    Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[0];
```

Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:179]:
```suggestion
                    Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[0];
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 25]
```diff
@@ -115,22 +138,74 @@ void Maestro::Evolve() {
...
+                    Real y = static_cast<Real>(j) * dx[1] + prob_lo[1];
+                    Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:181]:
```suggestion
#if (AMREX_SPACEDIM==3)
                    Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
#endif
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 26]
```diff
@@ -115,22 +138,74 @@ void Maestro::Evolve() {
...
+
+#if AMREX_SPACEDIM == 2
+                    vmac_arr(i,j,k) = std::pow(std::sin(M_PI*y), 2) * std::sin(2.0_rt*M_PI*x);
+#else
+
+                    vmac(i,j,k) =
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:187]:
```suggestion
                    vmac_arr(i,j,k) =
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 27]
```diff
@@ -115,22 +138,74 @@ void Maestro::Evolve() {
...
+                });
+
+#if AMREX_SPACEDIM == 3
+                // z-velocity  (x and y are centers, z are edges)
+
+                amrex::ParallelFor(mfi.nodaltilebox(2),
+                [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                {
+
+                    Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:200]:
```suggestion
                    Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[0];
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 28]
```diff
@@ -212,19 +295,128 @@ void Maestro::Evolve() {
...
 
-                add_grad_scalar(ARLIM_3D(lo), ARLIM_3D(hi),
-                                BL_TO_FORTRAN_FAB(gphi_mf[mfi]),
-                                BL_TO_FORTRAN_FAB(umid_mf[mfi]),
-                                phys_bc.dataPtr(), ZFILL(dx));
+                    amrex::ParallelFor(tilebox,
+                    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                    {
+
+                        Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:330]:
```suggestion
                        Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[0];
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 29]
```diff
@@ -212,19 +295,128 @@ void Maestro::Evolve() {
...
+                        Real y = (static_cast<Real>(j)+0.5_rt) * dx[1] + prob_lo[1];
+#if AMREX_SPACEDIM == 3
+                        Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
+#endif
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:334]:
```suggestion
#if (AMREX_SPACEDIM == 3)
                        Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
#endif
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 30]
```diff
@@ -212,19 +295,128 @@ void Maestro::Evolve() {
...
+
+                    amrex::ParallelFor(gbox,
+                    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                    {
+
+                        Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
+                        Real y = (static_cast<Real>(j)+0.5_rt) * dx[1] + prob_lo[1];
+#if AMREX_SPACEDIM == 3
+                        Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
+#endif
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:381]:
```suggestion
#if (AMREX_SPACEDIM == 3)
                        Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
#endif
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 31]
```diff
@@ -212,19 +295,128 @@ void Maestro::Evolve() {
...
+                    Elixir elix_phi = phi.elixir();
+                    auto phi_arr = phi.array();
+
+                    // now we can explicitly difference phi
+
+                    amrex::ParallelFor(gbox,
+                    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                    {
+
+                        Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:377]:
```suggestion
                        Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[0];
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 32]
```diff
@@ -212,19 +295,128 @@ void Maestro::Evolve() {
...
+
+                    Array4<Real> const gphi_arr = gphi_mf.array(mfi);
+                    Array4<Real> const umid_arr = umid_mf.array(mfi);
+
+
+                    amrex::ParallelFor(tilebox,
+                    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                    {
+
+                        Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:400]:
```suggestion
                        Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[0];
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 33]
```diff
@@ -212,19 +295,128 @@ void Maestro::Evolve() {
...
+                        Real y = (static_cast<Real>(j)+0.5_rt) * dx[1] + prob_lo[1];
+                        Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:402]:
```suggestion
#if (AMREX_SPACEDIM==3)
                        Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
#endif
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 34]
```diff
@@ -260,31 +452,220 @@ void Maestro::Evolve() {
...
+                    // x-velocity  (x are edges, y and z are centers)
+
+                    Array4<Real> const gphix_arr = gphix_mac_mf.array(mfi);
+                    Array4<Real> const umac_arr = umac_mid_mf.array(mfi);
+
+                    amrex::ParallelFor(mfi.nodaltilebox(0),
+                    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                    {
+
+                        Real x = static_cast<Real>(i) * dx[0] + prob_lo[2];
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:493]:
```suggestion
                        Real x = static_cast<Real>(i) * dx[0] + prob_lo[0];
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 35]
```diff
@@ -260,31 +452,220 @@ void Maestro::Evolve() {
...
+                        Real y = (static_cast<Real>(j)+0.5_rt) * dx[1] + prob_lo[1];
+                        Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:495]:
```suggestion
#if (AMREX_SPACEDIM==3)
                        Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
#endif
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 36]
```diff
@@ -212,19 +297,128 @@ void Maestro::Evolve() {
...
+                    ) {
+
+                    Array4<Real> const gphi_arr = gphi_mf.array(mfi);
+                    Array4<Real> const umid_arr = umid_mf.array(mfi);
+
+                    amrex::ParallelFor(tilebox,
+                    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                    {
+
+                        Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:330]:
```suggestion
                        Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[0];
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 37]
```diff
@@ -115,22 +140,74 @@ void Maestro::Evolve() {
...
+
+                // x-velocity  (x are edges, y and z are centers)
+
+                amrex::ParallelFor(mfi.nodaltilebox(0),
+                [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                {
+
+                    Real x = static_cast<Real>(i) * dx[0] + prob_lo[0];
+                    Real y = (static_cast<Real>(j)+0.5_rt) * dx[1] + prob_lo[1];
+                    Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:162]:
```suggestion
#if (AMREX_SPACEDIM==3)
                    Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
#endif
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 38]
```diff
@@ -115,22 +140,74 @@ void Maestro::Evolve() {
...
+#endif
+                });
+
+                // y-velocity  (x and z are centers, y are edges)
+
+                amrex::ParallelFor(mfi.nodaltilebox(1),
+                [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                {
+
+                    Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:179]:
```suggestion
                    Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[0];
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 39]
```diff
@@ -115,22 +140,74 @@ void Maestro::Evolve() {
...
+                    Real y = static_cast<Real>(j) * dx[1] + prob_lo[1];
+                    Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:181]:
```suggestion
#if (AMREX_SPACEDIM==3)
                    Real z = (static_cast<Real>(k)+0.5_rt) * dx[2] + prob_lo[2];
#endif
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 40]
```diff
@@ -115,22 +140,74 @@ void Maestro::Evolve() {
...
+
+#if AMREX_SPACEDIM == 2
+                    vmac_arr(i,j,k) = std::pow(std::sin(M_PI*y), 2) * std::sin(2.0_rt*M_PI*x);
+#else
+
+                    vmac(i,j,k) =
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:187]:
```suggestion
                    vmac_arr(i,j,k) =
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 41]
```diff
@@ -115,22 +140,74 @@ void Maestro::Evolve() {
...
+                });
+
+#if AMREX_SPACEDIM == 3
+                // z-velocity  (x and y are centers, z are edges)
+
+                amrex::ParallelFor(mfi.nodaltilebox(2),
+                [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k)
+                {
+
+                    Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[2];
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:200]:
```suggestion
                    Real x = (static_cast<Real>(i)+0.5_rt) * dx[0] + prob_lo[0];
```

File: Exec/unit_tests/test_projection/MaestroEvolve.cpp [hunk 42]
```diff
@@ -115,22 +140,80 @@ void Maestro::Evolve() {
...
+                    vmac_arr(i,j,k) = std::pow(std::sin(M_PI*y), 2) * std::sin(2.0_rt*M_PI*x);
+#else
+
+                    vmac_arr(i,j,k) =
+                        2.0_rt * M_PI * std::sin(4.0_rt*M_PI*y) * std::cos(2.0_rt*M_PI*z) -
+                        4.0_rt * M_PI * std::cos(4.0_rt*M_PI*x) * std::sin(2.0_rt*M_PI*y);
+#endif
+                });
+
+#if AMREX_SPACEDIM == 3
```
Comment (ajnonaka) [inline Exec/unit_tests/test_projection/MaestroEvolve.cpp:197]:
```suggestion
#if (AMREX_SPACEDIM == 3)
```

**Review by doreenfan — APPROVED** [review]

**Pull request merged**